src/dbus/server/dbus_thread_object.cpp

```diff
@@ -660,14 +660,19 @@ otError DBusThreadObject::GetLinkModeHandler(DBusMessageIter &aIter)
 
 otError DBusThreadObject::GetDeviceRoleHandler(DBusMessageIter &aIter)
 {
-    auto         threadHelper = mHost->GetThreadHelper();
-    otDeviceRole role         = otThreadGetDeviceRole(threadHelper->GetInstance());
-    std::string  roleName     = GetDeviceRoleName(role);
-    otError      error        = OT_ERROR_NONE;
+    otError error = OT_ERROR_NONE;
 
-    VerifyOrExit(DBusMessageEncodeToVariant(&aIter, roleName) == OTBR_ERROR_NONE, error = OT_ERROR_INVALID_ARGS);
+    mHost->GetDeviceRole([&aIter, &error](otError aError, otDeviceRole aRole) mutable {
```

Comment:
Per offline discussion with @Irving-cl, this piece of code only works in RCP mode so it may not be very useful to refactor it in this way. Ideally the handler will handle both RCP and NCP mode.

Reply:
Reverted the change.